src/main/xerus/monstercat/tabs/TabCatalog.kt

```diff
@@ -56,11 +60,90 @@ class TabCatalog : TableTab() {
 			it.next(); Settings.VISIBLECATALOGCOLUMNS.putMulti(*it.addedSubList.map { it.text }.toTypedArray())
 		})
 		table.setOnMouseClicked { me ->
-			if(me.clickCount == 2) {
+			if(me.clickCount == 2 && me.button == MouseButton.PRIMARY) {
 				val selected = table.selectionModel.selectedItem ?: return@setOnMouseClicked
 				Player.play(selected[cols.findUnsafe("Track")].trim(), selected[cols.findUnsafe("Artist")])
+			}else if(me.clickCount == 1 && me.button == MouseButton.MIDDLE){
+				val selected = table.selectionModel.selectedItem ?: return@setOnMouseClicked
+				GlobalScope.launch {
+					val track = APIUtils.find(selected[cols.findUnsafe("Track")].trim(), selected[cols.findUnsafe("Artist")])
+					if (track != null) Playlist.add(track)
+					else monsterUtilities.showMessage("The requested song could not be found.", "Cannot add to playlist", Alert.AlertType.WARNING)
+				}
+			}
+		}
+		table.setOnKeyPressed {
+			if (it.code == KeyCode.ENTER){
+				val selected = table.selectionModel.selectedItems
+				GlobalScope.launch {
+					Playlist.clear()
+					val tracklist = arrayListOf<Track>()
+					selected.forEach { item ->
+						val track = APIUtils.find(item[cols.findUnsafe("Track")].trim(), item[cols.findUnsafe("Artist")])
+						if (track != null) tracklist.add(track)
+						else logger.warn("Failed matching song ${item[cols.findUnsafe("Artist")]} - ${item[cols.findUnsafe("Track")].trim()} while adding it to playlist")
+					}
+					Player.playTracks(tracklist)
+				}
+			}else if(it.code == KeyCode.PLUS || it.code == KeyCode.ADD){
+				val selected = table.selectionModel.selectedItems
+				GlobalScope.launch {
+					selected.forEach { item ->
+						val track = APIUtils.find(item[cols.findUnsafe("Track")].trim(), item[cols.findUnsafe("Artist")])
+						if (track != null) Playlist.add(track)
+						else logger.warn("Failed matching song ${item[cols.findUnsafe("Artist")]} - ${item[cols.findUnsafe("Track")].trim()} while adding it to playlist")
+					}
+				}
+			}
+		}
+		
+		table.selectionModel.selectionMode = SelectionMode.MULTIPLE
+		val rightClickMenu = ContextMenu()
+		val item1 = MenuItem("Play") {
+			val selected = table.selectionModel.selectedItems
+			GlobalScope.launch {
+				Playlist.clear()
+				val tracklist = arrayListOf<Track>()
+				selected.forEach { item ->
+					val track = APIUtils.find(item[cols.findUnsafe("Track")].trim(), item[cols.findUnsafe("Artist")])
+					if (track != null) tracklist.add(track)
+					else logger.warn("Failed matching song ${item[cols.findUnsafe("Artist")]} - ${item[cols.findUnsafe("Track")].trim()} while adding it to playlist")
+				}
+				Player.playTracks(tracklist)
+			}
+		}
+		val item2 = MenuItem("Add to playlist") {
```

Comment:
Far too much repetition in this file

Reply:
Refactored in 602ac9706f2b48ed0a1a3c0835892da026a49b42